ARCH-115        add execution mode on Api definition

guillaumelamirand opened the pull request:
**Description**

Add execution mode to api definition and all related object used by both management api or the gateway.
Unit tests have been created for it.
Postman collections have been updated to test it


<!-- UI placeholder -->
🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.blob.core.windows.net/prototype-reactive-execution-mode/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)
<!-- UI placeholder end -->

Commit:
feat: add execution mode on Api definition

jhaeyaert commented on gravitee-apim-definition/gravitee-apim-definition-jackson/src/test/java/io/gravitee/definition/jackson/api/ApiDeserializerTest.java:
```diff
@@ -634,4 +635,25 @@ public void definition_v2_withPath() throws Exception {
     }
+
+    @Test
+    public void should_default_definition_executionMode_equal_v3_when_json_contains_null() throws Exception {
```
Can you switch to camelCase plz ?

guillaumelamirand replied:
yes, forget about this one.

jhaeyaert commented on gravitee-apim-rest-api/gravitee-apim-rest-api-model/src/main/java/io/gravitee/rest/api/model/api/ApiEntity.java:
```diff
@@ -62,6 +62,11 @@ public class ApiEntity implements Indexable {
     )
     private String description;
 
+    @Schema(description = "Api's execution mode. Define if the execution mode should use v3 or jupiter mode.", example = "v3")
```
"Define**s** if the execution mode should..." ?
Should we add a little description about what "jupiter" mode is for a better understanding ? 
Ex: "Jupiter execution mode allows strict respect of the policy execution order, as it appears it the policy studio whereas, in "v3" mode, execution order may differ depending on policy REQUEST_CONTENT or RESPONSE_CONTENT scope.

guillaumelamirand replied:
This annotation on other attribute is not used for documentation and contains basic information. not sure is the good place for this kind of details (or other need way more details also)

jhaeyaert commented on gravitee-apim-rest-api/gravitee-apim-rest-api-model/src/main/java/io/gravitee/rest/api/model/api/NewApiEntity.java:
```diff
@@ -47,6 +49,11 @@ public class NewApiEntity {
+    @Schema(description = "Api's execution mode. Define if the execution mode should use v3 or jupiter mode.", example = "v3")
+    @DeploymentRequired
+    @JsonProperty(value = "execution_mode")
+    private ExecutionMode executionMode;
```
Not sure we should require this information when creating a new api. Maybe we could let the service layer calculate the execution mode by itself instead of making it mandatory (ex: based on a feature flipping or any global configuration) ? WDYT ?

jhaeyaert commented on gravitee-apim-rest-api/gravitee-apim-rest-api-model/src/main/java/io/gravitee/rest/api/model/api/ApiListItem.java:
```diff
@@ -46,6 +48,10 @@ public class ApiListItem {
 
+    @Schema(description = "Api's execution mode. Define if the execution mode should use v3 or jupiter mode.", example = "v3")
+    @JsonProperty(value = "execution_mode")
+    private ExecutionMode executionMode;
```
I'm asking myself if it is necessary to add the executionMode on the `ApiListItem` entity 🧐
I'm not sure it will be a really useful information for api listing use cases. I'm also thinking about the future versions where the executionMode will certainly have no longer make sense and become really useless (and will create a breaking change if we want to remove it).

Do you think we can avoid it here ?

jhaeyaert commented on gravitee-apim-rest-api/gravitee-apim-rest-api-model/src/main/java/io/gravitee/rest/api/model/api/ApiQuery.java:
```diff
@@ -33,6 +34,7 @@ public class ApiQuery {
     private String label;
     private String state;
     private Visibility visibility;
+    private ExecutionMode executionMode;
```
Is there a utility to search by `executionMode` ? I'm not sure we have identify any use case yet. Do you think it is really necessary ?

Commit:
fix: rename unit test to camel case

Commit:
fix: remove execution mode from new entity, and list item

Commit:
feat: add new settings to enable global jupiter mode and set when to use it by default

update console settings to retrieve this settings for the UI
implement dedicated service to retrieve the execution mode from api definition

jhaeyaert commented on gravitee-apim-console-webui/src/management/api/creation/steps/api-creation.controller.ts:
```diff
@@ -13,15 +13,15 @@
  */
-import { StateService } from '@uirouter/core';
-import { IPromise } from 'angular';
+import {StateService} from '@uirouter/core';
```
It lacks a small touch of prettier here ;-)

jhaeyaert commented on gravitee-apim-console-webui/src/management/api/creation/steps/api-creation.controller.ts:
```diff
@@ -109,9 +110,16 @@ class ApiCreationController {
+      groups: []
     };
 
+    console.log(this.Constants.org.settings.jupiterMode)
```
I think you forgot to disable your debug mode 😜😂

guillaumelamirand replied:
oups !

jhaeyaert commented on gravitee-apim-definition/gravitee-apim-definition-jackson/src/test/java/io/gravitee/definition/jackson/api/ApiSerializerTest.java:
```diff
@@ -662,4 +662,52 @@ public void definition_kafka_endpoint() throws Exception {
     }
+
+    @Test
+    public void should_default_definition_executionMode_equal_v3_when_api_contains_null() throws Exception {
```
camelCase ?

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-distribution/src/main/resources/config/gravitee.yml:
```diff
@@ -406,8 +406,13 @@ ds:
 
-# Encrypt API properties using this secret
 api:
+  # Jupiter's execution mode allows strict respect of the policy execution order, as it appears it the policy studio whereas, in "v3" mode, execution order may differ depending on policy REQUEST_CONTENT or RESPONSE_CONTENT scope.
```
Little typo "as it appears it the policy studio" -> "as it appears **in** the policy studio"

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-distribution/src/main/resources/config/gravitee.yml:
```diff
@@ -406,8 +406,13 @@ ds:
+  # jupiterMode:
+    # Enable jupiter mode
+    # Default : false
+    # enabled: false
```
Can you replace by:

```yaml
  # jupiterMode:
    # Enable jupiter mode, default is false
    # enabled: false
```

I just want to avoid confusion regarding the `# Default : false` which could be miss interpreted as a configuration entry.

guillaumelamirand replied:
I was thinking the same but this is how default is explained on some others attributes such as unknown-expire-after

jhaeyaert commented on gravitee-apim-rest-api/gravitee-apim-rest-api-model/src/main/java/io/gravitee/rest/api/model/parameters/Key.java:
```diff
@@ -301,7 +301,9 @@ public enum Key {
-    CONSOLE_PATH_BASED_API_CREATION_ENABLED("console.pathBasedApiCreation.enabled", "false", Set.of(ORGANIZATION, SYSTEM));
+    CONSOLE_PATH_BASED_API_CREATION_ENABLED("console.pathBasedApiCreation.enabled", "false", Set.of(ORGANIZATION, SYSTEM)),
+    JUPITER_MODE_ENABLED("api.jupiterMode.enabled", "false", Set.of(ENVIRONMENT, SYSTEM)),
+    JUPITER_MODE_DEFAULT("api.jupiterMode.default", "always", Set.of(ENVIRONMENT, SYSTEM));
```
As it is a global configuration, does it make sense to scope it to the `ENVIRONMENT` ? Does `SYSTEM` only is enough ?

guillaumelamirand replied:
I need to try again but it wasn't working as expected without ENVIRONMENT.

guillaumelamirand replied:
Ok after a test, it is now working as expected only with SYSTEM. I probably mix issue.

jhaeyaert commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/JupiterModeServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+                }
+            }
+        }
+        return ExecutionMode.V3;
```
Can be simplified with a more readable version:

```java
if(enabled && (defaultMode == ALWAYS || (apiDefinition == null && defaultMode == CREATION_ONLY))) {
    return ExecutionMode.JUPITER;
}

return ExecutionMode.V3
```

WDYT ?

guillaumelamirand replied:
it is a point of view, I don't think it is more readable with one big IF and || and &&

guillaumelamirand replied:
I've updated as suggested to simplify the condition however the formatter avoid me to make it more readable...I put a small comment.

jhaeyaert commented on gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/ApiReactorHandlerFactory.java:
```diff
@@ -157,7 +158,11 @@ public ReactorHandler create(Api api) {
 
-                if (api.getExecutionMode() == null || api.getExecutionMode() == ExecutionMode.V3) {
+                if (
+                    !configuration.getProperty(API_JUPITER_MODE_ENABLED, Boolean.class) ||
```
Add default value to false to avoid possible NPE: `!configuration.getProperty(API_JUPITER_MODE_ENABLED, Boolean.class, false)`

jhaeyaert commented on gravitee-apim-rest-api/gravitee-apim-rest-api-standalone/gravitee-apim-rest-api-standalone-distribution/src/main/resources/config/gravitee.yml:
```diff
@@ -546,8 +546,16 @@ notifiers:
+  # Jupiter's execution mode allows strict respect of the policy execution order, as it appears it the policy studio whereas, in "v3" mode, execution order may differ depending on policy REQUEST_CONTENT or RESPONSE_CONTENT scope.
+  # jupiterMode:
+    # Enable jupiter mode
+    # Default : false
```
Same here

jhaeyaert commented on gravitee-apim-rest-api/gravitee-apim-rest-api-standalone/gravitee-apim-rest-api-standalone-distribution/src/main/resources/config/gravitee.yml:
```diff
@@ -546,8 +546,16 @@ notifiers:
+    # Default : false
+    # enabled: false
+    # Allow to use the jupiter mode by default when no execution mode is specified on an api.
+    # Default : always
```
And here :-)

jhaeyaert reviewed (changes requested):
Tested locally and works as expected on the different cases, well done 👍 
Only few comments here and there
